web/modules/custom/joinup_user/joinup_user.module

```diff
@@ -329,8 +441,42 @@ function joinup_user_entity_base_field_info(EntityTypeInterface $entity_type) {
  * Implements hook_ENTITY_TYPE_presave() for OG Membership entities.
  */
 function joinup_user_og_membership_presave(EntityInterface $entity) {
-  // When a membership changes, clear the user's cache so that the related
-  // content on the user profile page matches the new group memberships.
+  joinup_user_invalidate_user_cache_tags($entity);
+
+  if ($entity->getEntityTypeId() === 'og_membership') {
+    /** @var \Drupal\user\Entity\User $account */
+    $account = $entity->getUser();
+
+    $params = [
+      'account' => $account,
+      'membership' => $entity,
+    ];
+
+    $params['account'] = $account;
+    \Drupal::service('plugin.manager.mail')->mail(
+      'joinup_user',
+      'og_roles_changed',
+      $account->getEmail(),
+      $account->getPreferredLangcode(),
+      $params
+    );
+  }
+}
+
+/**
+ * Implements hook_ENTITY_TYPE_predelete() for OG Membership entities.
+ */
+function joinup_user_og_membership_predelete(EntityInterface $entity) {
+  joinup_user_invalidate_user_cache_tags($entity);
+}
+
+/**
+ * Flush the caches of the user that belongs to a OG membership.
+ *
+ * @param \Drupal\Core\Entity\EntityInterface $entity
+ *   The OG membership entity.
+ */
+function joinup_user_invalidate_user_cache_tags(EntityInterface $entity) {
   if ($entity->getEntityTypeId() === 'og_membership') {
```

Comment:
If you type hint on `OgMembershipInterface` instead of on the generic `EntityInterface` then you don't need this switch.